schema/encoding/jsonschema.go

```diff
@@ -0,0 +1,110 @@
+	"strconv"
+	"strings"
+)
+
+// JSONSchemaPropertyBase common Embedded Values
+type JSONSchemaPropertyBase struct {
+	Title string `json:"title"`
+	Type  string `json:"type"`
+}
+
+// JSONSchema represents a standard JSONSchema
+type JSONSchema struct {
+	JSONSchemaPropertyBase
+	Properties map[string]map[string]interface{} `json:"properties"`
+}
+
+// BoundariesToJSONSchema render boundaries to JSON
+func BoundariesToJSONSchema(b *schema.Boundaries) string {
+	if b == nil {
+		return ""
+	}
+	return fmt.Sprintf("%q: %s, %q: %s",
+		"minimum", strconv.FormatFloat(b.Min, 'E', -1, 64),
+		"maximum", strconv.FormatFloat(b.Max, 'E', -1, 64))
+}
+
+// ValidatorToJSONSchema takes a validator and renders to JSON
+func ValidatorToJSONSchema(v schema.FieldValidator) string {
+	if v == nil {
+		return ""
+	}
+	var result string
+	switch t := v.(type) {
+	case *schema.String:
+		result += fmt.Sprintf("%q: %q, ", "type", "string")
+		if t.Regexp != "" {
+			result += fmt.Sprintf("%q: %q, ", "pattern", t.Regexp)
+		}
+		if len(t.Allowed) > 0 {
+			var allowed []string
+			for _, value := range t.Allowed {
+				allowed = append(allowed, fmt.Sprintf("%q", value))
+			}
+			result += fmt.Sprintf("%q: [ %s ],", "oneOf", strings.Join(allowed, ","))
+		}
+		if t.MinLen > 0 {
+			result += fmt.Sprintf("%q: %s, ", "minLength", strconv.FormatInt(int64(t.MinLen), 10))
+		}
+		if t.MaxLen > 0 {
+			result += fmt.Sprintf("%q: %s, ", "maxLength", strconv.FormatInt(int64(t.MaxLen), 10))
+		}
+	case *schema.Integer:
+		result += fmt.Sprintf("%q: %q, ", "type", "number")
+		if len(t.Allowed) > 0 {
+			var allowed []string
+			for _, value := range t.Allowed {
+				allowed = append(allowed, strconv.FormatInt(int64(value), 10))
+			}
+			result += fmt.Sprintf("%q: [ %s ],", "oneOf", strings.Join(allowed, ","))
+		}
+		result += fmt.Sprintf("%s,  ", BoundariesToJSONSchema(t.Boundaries))
+	case *schema.Array:
+		result += fmt.Sprintf("%q: %q, ", "type", "array")
+		if t.ValuesValidator != nil {
+			result += fmt.Sprintf("%q: %s, ", "items", ValidatorToJSONSchema(t.ValuesValidator))
+		}
+	case *schema.Object:
+		result += fmt.Sprintf("%s, ", SchemaToJSONSchema(t.Schema))
+	case *schema.Time:
+		result += fmt.Sprintf("%q: %q, ", "type", "string")
+	default:
+	}
+	result = strings.Trim(result, ", ")
+	return result
+}
+
+// SchemaToJSONSchema helper
+func SchemaToJSONSchema(s *schema.Schema) string {
+	result := "{"
+	if s.Description != "" {
+		result += fmt.Sprintf("%q: %q, ", "title", s.Description)
+	}
+	result += fmt.Sprint(`"type": "object",`)
+	result += fmt.Sprint(`"properties": {`)
+	var required []string
+	for key, field := range s.Fields {
+		result += fmt.Sprintf("%q: {", key)
+		if field.Description != "" {
+			result += fmt.Sprintf("%q: %q, ", "description", field.Description)
+		}
+		if field.Required {
+			required = append(required, fmt.Sprintf(`"%s"`, key))
+		}
+		if field.ReadOnly {
+			result += fmt.Sprintf("%q: %t, ", "readonly", field.ReadOnly)
```

Comment:
readonly  is not really defined in  Vanilla JSON Schema. It's one of the very few things added by Swagger (a.k.a. Open API Specification) actually.